### DISCUSSION TITLE
Equal stresses across gas-liquid interface using INS

### DISCUSSION THREAD

**Discussion by keniley1** [idaholab/moose#17220] [Q&A Modules: Navier-Stokes] [answered]
Hello all,

I'm trying to model an argon jet interacting with a liquid water surface. The setup is essentially a cylindrical tube with a hot tungsten needle at the center, with the tip of the needle reaching temperatures around 2000-2500 K. The tip of the needle is suspended over a reservoir of initially stationary liquid water at room temperature. The flow at the inlet is 8.8 cm/s and the inner radius of the inlet tube is 2 mm, so we believe laminar flow is a very safe assumption. I was able to set up a four region problem (gas phase, heated needle, gas flow tube surrounding the needle, and water) including the incompressible Navier-Stokes equations in the gas and liquid and heat transfer in all 4 regions (see the attached image). I'm using the RZ INS equations in this case.

![needle_airflow](https://user-images.githubusercontent.com/16326064/109886522-cb47cc80-7c45-11eb-87d6-bee408e80440.png)

I'd like to see how the flow induced in the liquid water will affect the temperature distribution and, eventually, evaporation from the liquid.  Right now I'm simply setting the radial component of the velocities equal to each other at the interface and setting the z-component to zero, but I believe a better no slip boundary condition would be equal stress terms at the gas-liquid interface based on this paper: https://iopscience.iop.org/article/10.1088/0022-3727/48/42/424007/meta 

Following the Moose INS paper published a couple years ago (https://www.sciencedirect.com/science/article/pii/S0965997817310591) I think this would require an Interface Kernel with the (-p **I** + \mu \nabla \vec{u}) terms set equal to each other on either side of the interface. I couldn't find anything like this in the INS tests.  I guess this can be split into three questions: 

1. Am I interpreting an "equal stress" interfacial condition correctly? 
2. Does anything like this exist in Moose? 
3. If not, how difficult would it be to implement? 

EDIT: 
I was able to run these simulations by setting the y component of the velocity equal to zero at the water interface and setting no BC on the x component of velocity. (Previously I was setting vel_x equal at the interface.) I applied no interfacial BC for the pressure or temperature. The idea was that setting no BC on an interface will enforce natural continuity conditions since the variables vel_x, vel_y, p, and T exist in both blocks. I think that's true with temperature, but I have no idea what setting no interfacial condition means for velocity or pressure.

The results seem much more reasonable than before and qualitatively the velocity profiles in the gas and liquid look like I would expect, but I'm still not confident that I am enforcing the correct boundary conditions here. What would a continuity of normal and shear stress interfacial condition look like in Moose?

**Comment by GiudGiud** [comment, marked as answer]
Hello

Sorry we missed this. Your edit made it pop back up on our radar. 

A few questions about what you are trying to do:
- are you going to track the two species (argon and H2O) in the gaseous domain?
- are you still trying to model the movement of the water-gas interface? Or are you satisfied with vy=0 ?

I dont think anyone has done this in MOOSE yet, but the community is large so I may just be unaware. 
Making an interface kernel that sets the terms to be equal would not be hard to do.

Best,
Guillaume

**Reply by keniley1** [reply]
Thanks for the reply!

> are you going to track the two species (argon and H2O) in the gaseous domain?

No, I am only interested in the velocity and temperature profiles in both regions at the moment.

> are you still trying to model the movement of the water-gas interface? Or are you satisfied with vy=0 ?

The inlet speed is quite low and the flow is laminar so I am content with a static surface, vy = 0. I would definitely be interested to see how I could deal with surface deformation in the future though. 

-Shane

**Reply by GiudGiud** [reply]
Hi Shane

Ok single species should be easier. We are adding multi-species transport to our finite volume Navier Stokes implementation, but it's not quite ready yet. We won't be developing species tracking in FE afaik.

Surface deformation is actually quite tricky. I don't think it's impossible, but we have added capabilities for fluid-structure interaction, not fluid-fluid interaction.

I think an interface kernel is the right object to model an equal stress boundary, and it should not be too hard to do. Whether it's the correct thing to do, I think it makes physical sense to have continuity of the stresses.
You already did an integration by parts on this term and are looking at the surface contribution right?

What family of variables are you using? Are you sure you want no-slip BCs (=continuous velocities)? It seems to me like the gas would flow faster than the fluid. 

Guillaume

**Reply by keniley1** [reply]
> You already did an integration by parts on this term and are looking at the surface contribution right?

Pressure is integrated by parts, if that's what you mean. All I did was apply the INS kernels in RZ coordinates (INSMassRZ, INSMomentumLaplaceFormRZ, and INSTemperature). What I'm trying to do is set the appropriate boundary condition for two immiscible liquids. That should be continuity of normal and shear stresses across the interface but I have no idea how to actually create that condition. (I agree that it would be an interface kernel, though.)

> What family of variables are you using? 

I am using second order lagrange variables for the velocity and temperature and first order lagrange for pressure. integrate_p_by_parts is set to true. 

> Are you sure you want no-slip BCs (=continuous velocities)? It seems to me like the gas would flow faster than the fluid.

I am not applying no slip BCs on the interface anymore. Now only vel_y components are zero at the interface, but no condition is applied for vel_x. As you say, the gas should be moving quite a bit faster than the liquid. That was actually the point that led to this question! 

Right now I have variables p, vel_x, and vel_y, and T which all exist in both gas and liquid blocks, and I set no interfacial conditions for vel_x or pressure. With these conditions the results look _qualitatively_ good -- the gas is moving much faster than the liquid and a convective cell forms in the water. Even so, I am just not convinced that continuity of normal and shear stress is being properly enforced if I set no interface condition. 

So I guess my question now is: what would a stress continuity interface kernel look like?  I believe the stress component is 

`pI - \mu \grad u`

but this forms a tensor, and I cannot figure out how that can be applied in an interface kernel.

I've created a branch with a couple input files to demonstrate what I'm doing, if that would help. I'm running these in another Moose app (zapdos), but they only use Moose kernels so the input files should run anywhere with heat conduction and navier stokes modules. They are in zapdos/problems/cfd_tests in this branch: 

https://github.com/keniley1/zapdos/tree/cfd_interface_tests

The pin_water.i file is the geometry I presented above. The two_fluid_pipe.i file is a simpler steady problem consisting of two immiscible fluids driven by a pressure differential through a pipe.

**Reply by GiudGiud** [reply]
I think using this continuous variable family defined over both domains will impose continuity of velocity and pressure, which is not what you want.

I think you should have two sets of variables defined on each domain, with pretty much the same kernels but different material properties. 

The connection between the two domain could then be done with an interface kernel. The interface kernel would impose the stress continuity as a constraint, normal and tangeantial components, with a Lagrange multiplier (one for each maybe?).

What do you think about this?
I'm not very familiar with FE interface kernels. We'll have to tag more expert people later on

**Reply by keniley1** [reply]
I'm not sure how to apply Lagrange multipliers here or how it would apply to stress conservation, but I'm willing to learn and try it out!

At any rate, I decided to try to go forward with an InterfaceKernel. I just brute forced it so the code is hideous, but I basically just computed each component of the 2D stress tensor by hand and set the "Element" stress component equal to the "Neighbor" stress component (and vice-versa). This requires applying only two interface kernels, one for each velocity component. I really didn't think this would work because I cannot wrap my head around the idea of a tensor interface condition, but to my surprise, this gave me the correct analytical solution for the pipe flow problem I uploaded. I also needed to use MatchedValueBC to force the velocities to be equal at the interface. I believe that makes sense. 

I've attached plots of (a) the 2D x-velocity profile, and (b) 1D x-velocity profiles in the two regions. In (b), the black curve is the numerical solution and red dotted curve is the analytical one. I got the analytical solution from Bird's book _Transport Phenomena_ (2nd edition, section 2.5, "Flow of Two Adjacent Immiscible Fluids").  The interface kernel is called "StressContinuity" and is in the branch I linked to previously.  I used an ADInterfaceKernel so I didn't have to bother with jacobians. The percent difference from the analytical solution is basically 0. 

**Parameters:**
_Pipe length_ - 1 
_Pipe width_ - 0.25
_pressure difference_ - 10

_Block 0 (bottom)_ -  \mu = 3, \rho = 2
_Block 1 (top)_ - \mu = 1, \rho = 1

Even more surprising: this is the exact same solution that I get when I keep the single continuous variable across both regions. This makes no sense to me because I totally agreed with you -- I didn't think using a continuous variable would apply the correct interfacial condition. Apparently it does. I have no idea how this works! 

![2d_plot](https://user-images.githubusercontent.com/16326064/120850401-00add000-c53d-11eb-9bec-56a22a4a0f18.png)
![pipe_vx_solution](https://user-images.githubusercontent.com/16326064/120850404-02779380-c53d-11eb-9cec-04008a5b495e.png)

**Reply by GiudGiud** [reply]
Well if you force variable continuity, and use the same kernels, and essentially have the same contributions to the residual from the interface kernels as you would from kernels (this part I am unsure, it feels like with variable continuity you may have done that), then it's re-assuring that you got the same solution with two variables defined on each part of the domain and with one variable defined on the whole domain. 

So the benchmark has continuous velocities. Do you think this should apply to the gas-liquid case too?

**Reply by keniley1** [reply]
I believe so, yes. It's certainly an approximation, but the flow in my problem is laminar, no surface deformation is observed in the experiment, and the gas is nonreactive (it's an argon environment over liquid water) so I think the immiscible fluid assumption still holds. The continuous velocities with equal stress tensors has also been used in a similar problems in the past (see [this paper](https://iopscience.iop.org/article/10.1088/0022-3727/48/42/424007/meta)), so there's a precedent. I guess this question is answered!  Thanks for the advice.

If there is any interest in having a stress continuity interface kernel in Moose I would not mind generalizing the one I made to 3D and submitting a PR. That said, I'm not sure if it has any use since it appears to be equivalent to using a single velocity vector.
